### PR TITLE
[Sparse] Add a graphSAGE example

### PR DESCRIPTION
**Pull request by xiangyuzhi**
## Description
A sparse graphSAGE example. 
Run example: 
```
$ python examples/sparse/graphsage.py
Training in gpu mode.
Loading data
Training...
[W TensorAdvancedIndexing.cpp:1615] Warning: scatter_reduce() is in beta and the API may change at any time. (function operator())
Epoch 00000 | Loss 2.0963 | Accuracy 0.7756 
Epoch 00001 | Loss 0.8108 | Accuracy 0.8443 
Epoch 00002 | Loss 0.6264 | Accuracy 0.8593 
Epoch 00003 | Loss 0.5662 | Accuracy 0.8675 
Epoch 00004 | Loss 0.5311 | Accuracy 0.8721 
Epoch 00005 | Loss 0.5047 | Accuracy 0.8744 
Epoch 00006 | Loss 0.4846 | Accuracy 0.8759 
Epoch 00007 | Loss 0.4766 | Accuracy 0.8756 
Epoch 00008 | Loss 0.4674 | Accuracy 0.8796 
Epoch 00009 | Loss 0.4528 | Accuracy 0.8799 
Testing...
Test accuracy 0.7216
```
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes

### CODE REVIEW

File: examples/sparse/graphsage.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        if self._aggre_type != "gcn":
+            nn.init.xavier_uniform_(self.fc_self.weight, gain=gain)
+        nn.init.xavier_uniform_(self.fc_neigh.weight, gain=gain)
+
+    def forward(self, A, feat):
+        feat_src = feat_dst = self.feat_drop(feat)
+        feat_dst = feat_src[: A.shape[0]]
+
+        h_self = feat_dst
+        lin_before_mp = self._in_src_feats > self._out_feats
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:129]: To simplify this example, you can always apply linear before message passing.

File: examples/sparse/graphsage.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        srcdata, dstdata = {}, {}
+        if self._aggre_type == "mean":
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:131]: Assume the aggregation type is 'mean' for simplification.

File: examples/sparse/graphsage.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+            srcdata["h"] = (
+                self.fc_neigh(feat_src) if lin_before_mp else feat_src
+            )
+            # Div by degree
+            D_hat = dglsp.diag(A_hat.sum(0)) ** -1
+            A_div = A_hat @ D_hat
+            dstdata["neigh"] = A_div.T @ srcdata["h"]
+            h_neigh = graph.dstdata["neigh"]
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:139]: Where is `graph` from?

File: examples/sparse/graphsage.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+    def __init__(
+        self,
+        in_feats,
+        out_feats,
+        aggregator_type="gcn",
+        feat_drop=0.0,
+        bias=True,
+        norm=None,
+        activation=None,
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:96]: These parameters are not needed.

File: examples/sparse/graphsage.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        acc = evaluate(A, features, labels, val_mask, model)
+        print(
+            f"Epoch {epoch:05d} | Loss {loss.item():.4f} | Accuracy {acc:.4f} "
+        )
+
+
+class SAGEConv(nn.Module):
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:83]: Put this class on top of `SAGE` for readability.

**Review by czkkkkkk — CHANGES_REQUESTED** [review]

File: examples/sparse/graphsage.py (hunk 3 above)
Comment (xiangyuzhi) [reply examples/sparse/graphsage.py]: It is a typo mistake, I have fixed this.

**Comment by xiangyuzhi** [comment]
@dgl-bot

**Comment by czkkkkkk** [comment]
This should serve as a sampling-based training example, illustrating how sparse APIs support sampling. Let's have a chat tomorrow to discuss how to implement these changes.

File: examples/sparse/graphsage.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        gain = nn.init.calculate_gain("relu")
+        nn.init.xavier_uniform_(self.fc_self.weight, gain=gain)
+        nn.init.xavier_uniform_(self.fc_neigh.weight, gain=gain)
+
+    def forward(self, A, feat):
+        # Remove duplicate edges.
+        A = A.coalesce()
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:59]: Is `A` coalesced after `compact()`?

File: examples/sparse/graphsage.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        feat_src = feat_dst = feat
+        feat_dst = feat_src[: A.shape[1]]
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:61]:
```suggestion
        feat_src = feat
        feat_dst = feat[: A.shape[1]]
```

File: examples/sparse/graphsage.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+    )
+    val_dataloader = torch.utils.data.DataLoader(val_idx, batch_size=1024)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    fanout = [10, 10, 10]
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (dst) in enumerate(train_dataloader):
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:197]:
```suggestion
        for it, (seeds) in enumerate(train_dataloader):
```

File: examples/sparse/graphsage.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+            # print(dst)
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:198]:
```suggestion
```

File: examples/sparse/graphsage.py [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    val_idx = dataset.val_idx.to(device)
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_idx, batch_size=1024, shuffle=True
+    )
+    val_dataloader = torch.utils.data.DataLoader(val_idx, batch_size=1024)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    fanout = [10, 10, 10]
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:193]:
```suggestion
    fanouts = [10, 10, 10]
```

File: examples/sparse/graphsage.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (dst) in enumerate(train_dataloader):
+            # print(dst)
+            src = dst
+            A_sample = []
+            for fout in fanout:
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:201]:
```suggestion
            for fanout in fanouts:
```

File: examples/sparse/graphsage.py [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+                # Sampling neighbor
+                mat = A.sample(1, fout, ids=src, replace=True)
+                # Compact the matrix
+                mat_cmp, src_idx = mat.compact(0)
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:205]:
```suggestion
                compacted_mat, row_ids = mat.compact(0)
```

File: examples/sparse/graphsage.py [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (dst) in enumerate(train_dataloader):
+            # print(dst)
+            src = dst
+            A_sample = []
+            for fout in fanout:
+                # Sampling neighbor
+                mat = A.sample(1, fout, ids=src, replace=True)
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:203]:
```suggestion
                sampled_mat = A.sample(1, fout, ids=src, replace=True)
```

File: examples/sparse/graphsage.py [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    fanout = [10, 10, 10]
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (dst) in enumerate(train_dataloader):
+            # print(dst)
+            src = dst
+            A_sample = []
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:200]:
```suggestion
            sampled_mats = []
```

File: examples/sparse/graphsage.py [hunk 15]
```diff
@@ -0,0 +1,283 @@
...
+            for fout in fanout:
+                # Sampling neighbor
+                mat = A.sample(1, fout, ids=src, replace=True)
+                # Compact the matrix
+                mat_cmp, src_idx = mat.compact(0)
+                A_sample.append(mat_cmp)
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:206]:
```suggestion
                sampled_mats.insert(0, compacted_mat)
```

File: examples/sparse/graphsage.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+                src = src_idx
+
+            A_sample.reverse()
+            csrc = src.to("cpu")
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:210]: Why move `src` and `dst` to CPU? The dataset should be moved to `device` before the index select.

File: examples/sparse/graphsage.py [hunk 17]
```diff
@@ -0,0 +1,283 @@
...
+                epoch, total_loss / (it + 1), acc.item()
+            )
+        )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="GraphSAGE")
+    parser.add_argument(
+        "--mode",
+        default="puregpu",
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:235]:
```suggestion
        default="gpu",
```

File: examples/sparse/graphsage.py [hunk 18]
```diff
@@ -0,0 +1,283 @@
...
+        choices=["cpu", "puregpu"],
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:236]: ditto

**Review by czkkkkkk — CHANGES_REQUESTED** [review]

File: examples/sparse/graphsage.py (hunk 6 above)
Comment (xiangyuzhi) [reply examples/sparse/graphsage.py:59]: No, it will occur an error without coalescing.

File: examples/sparse/graphsage.py (hunk 16 above)
Comment (xiangyuzhi) [reply examples/sparse/graphsage.py:210]: Ok.

File: examples/sparse/graphsage.py [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+    # Create sampler & dataloader.
+    train_idx = dataset.train_idx.to(device)
+    val_idx = dataset.val_idx.to(device)
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_idx, batch_size=1024, shuffle=True
+    )
+    val_dataloader = torch.utils.data.DataLoader(val_idx, batch_size=1024)
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:147]:
```suggestion
    optimizer = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
```

File: examples/sparse/graphsage.py [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+
+    fanouts = [10, 10, 10]
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (seeds) in enumerate(train_dataloader):
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:153]:
```suggestion
        for it, seeds in enumerate(train_dataloader):
```

File: examples/sparse/graphsage.py [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+                hidden_x = self.dropout(hidden_x)
+        return hidden_x
+
+
+def evaluate(model, A, dataloader, ndata, num_classes):
+    model.eval()
+    ys = []
+    y_hats = []
+    fanouts = [10, 10, 10]
+    for it, (seeds) in enumerate(dataloader):
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:104]:
```suggestion
    for it, seeds in enumerate(dataloader):
```

File: examples/sparse/graphsage.py [hunk 22]
```diff
@@ -0,0 +1,230 @@
...
+        with torch.no_grad():
+            src = seeds
+            sampled_mats = []
```
Comment (czkkkkkk) [inline examples/sparse/graphsage.py:107]: Please move L107-L114 to a separate function and highlight that it uses sparse API for subgraph sampling.

**Review by czkkkkkk — COMMENTED** [review]
LGTM

File: examples/sparse/graphsage.py (hunk 6 above)
Comment (czkkkkkk) [reply examples/sparse/graphsage.py:59]: Which operator caused the problem?

Comment (xiangyuzhi) [reply examples/sparse/graphsage.py:59]: Maybe because I set the parameter replace=True in compact(), I will remove it.

File: examples/sparse/graphsage.py
Comment (frozenbugs) [inline examples/sparse/graphsage.py:1]: examples/sparse/sampling.py

or 

examples/sparse/sampling/graphsage.py

File: examples/sparse/graphsage.py [hunk 23]
```diff
@@ -0,0 +1,230 @@
+"""
+This script trains and tests a GraphSAGE model based on the information of
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:2]: Please clarify what does this example demonstrate, emphasize the sparse sampling here.

File: examples/sparse/graphsage.py [hunk 24]
```diff
@@ -0,0 +1,230 @@
+"""
+This script trains and tests a GraphSAGE model based on the information of 
+a full graph.
+
+This flowchart describes the main functional sequence of the provided example.
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:5]: ditto, sampling is not highlighted in the flowchart.

File: examples/sparse/graphsage.py [hunk 25]
```diff
@@ -0,0 +1,230 @@
...
+        self.layers = nn.ModuleList()
+        # Three-layer GraphSAGE-gcn.
+        self.layers.append(SAGEConv(in_size, hid_size))
+        self.layers.append(SAGEConv(hid_size, hid_size))
+        self.layers.append(SAGEConv(hid_size, out_size))
+        self.dropout = nn.Dropout(0.5)
+        self.hid_size = hid_size
+        self.out_size = out_size
+
+    def forward(self, sampled_mats, x):
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:87]: rename all *mat to *matrix, unnecessary abbreviation hurts readability.

File: examples/sparse/graphsage.py [hunk 26]
```diff
@@ -0,0 +1,230 @@
...
+        self.fc_self = nn.Linear(self._in_dst_feats, out_feats, bias=True)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        gain = nn.init.calculate_gain("relu")
+        nn.init.xavier_uniform_(self.fc_self.weight, gain=gain)
+        nn.init.xavier_uniform_(self.fc_neigh.weight, gain=gain)
+
+    def forward(self, A, feat):
+        # Remove duplicate edges.
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:58]:
```suggestion
        # Remove duplicated edges.
```

File: examples/sparse/graphsage.py [hunk 27]
```diff
@@ -0,0 +1,230 @@
...
+    model.eval()
+    ys = []
+    y_hats = []
+    fanouts = [10, 10, 10]
+    for it, (seeds) in enumerate(dataloader):
+        with torch.no_grad():
+            src = seeds
+            sampled_mats = []
+            for fanout in fanouts:
+                # Sampling neighbor
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:109]:
```suggestion
                # Sample neighbors.
```

File: examples/sparse/graphsage.py [hunk 28]
```diff
@@ -0,0 +1,230 @@
...
+                sampled_mat = A.sample(1, fanout, ids=src, replace=True)
+                # Compact the matrix
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:111]:
```suggestion
                # Compact the sampled matrix.
```

File: examples/sparse/graphsage.py [hunk 29]
```diff
@@ -0,0 +1,230 @@
...
+
+    opt = torch.optim.Adam(model.parameters(), lr=1e-3, weight_decay=5e-4)
+
+    fanouts = [10, 10, 10]
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (seeds) in enumerate(train_dataloader):
+            src = seeds
+            sampled_mats = []
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:155]: The the train / validation and test code are quite similar from iterate the dataloader to generate sampled_mats and x, y. consider refactor to simplify the code.

File: examples/sparse/graphsage.py [hunk 30]
```diff
@@ -0,0 +1,230 @@
...
+        )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="GraphSAGE")
+    parser.add_argument(
+        "--mode",
+        default="gpu",
+        choices=["cpu", "gpu"],
+        help="Training mode. 'cpu' for CPU training, "
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:187]: nit: make full use of this line before change to new line.

File: examples/sparse/graphsage.py [hunk 31]
```diff
@@ -0,0 +1,230 @@
...
+        "'gpu' for pure-GPU training.",
+    )
+    args = parser.parse_args()
+    if not torch.cuda.is_available():
+        args.mode = "cpu"
+    print(f"Training in {args.mode} mode.")
+
+    #####################################################################
+    # (HIGHLIGHT) Node classification task is a supervise learning task
```
Comment (frozenbugs) [inline examples/sparse/graphsage.py:196]: Is this the HIGHLIGHT of the example? I think the HIGHLIGHT should be sampling with sparse matrix.

File: examples/sparse/sampling/graphsage.py [hunk 32]
```diff
@@ -0,0 +1,234 @@
+"""
+This script demonstrate how to use sparse to sample on graph and train model.
```
Comment (frozenbugs) [inline examples/sparse/sampling/graphsage.py:2]: how to use dgl sparse library to sample ...

File: examples/sparse/sampling/graphsage.py [hunk 33]
```diff
@@ -0,0 +1,234 @@
...
+    return MF.accuracy(
+        torch.cat(y_hats),
+        torch.cat(ys),
+        task="multiclass",
+        num_classes=num_classes,
+    )
+
+
+def validate(device, A, ndata, dataset, model, batch_size):
+    model.eval()
```
Comment (frozenbugs) [inline examples/sparse/sampling/graphsage.py:145]: model.eval() is  called in evaluate, redundant here.

**Review by frozenbugs — APPROVED** [review]

**Review by czkkkkkk — APPROVED** [review]
LGTM

**Pull request merged**